pwnlib/fmtstr.py

```diff
@@ -0,0 +1,92 @@
+import re
+import logging
+
+from pwnlib.util.cyclic import *
+from pwnlib.util.packing import *
+from pwnlib.util.fiddling import randoms
+from pwnlib.memleak import MemLeak
+from pwnlib.log import getLogger
+
+log = getLogger("pwnlib.fmtstr")
+
+class FmtStr(object):
+    def __init__(self, execute_fmt, offset = None, padlen = 0, numbwritten = 0):
+        self.execute_fmt = execute_fmt
+        self.offset = offset
+        self.padlen = padlen
+        self.numbwritten = numbwritten
+
+
+        if self.offset == None:
+            self.offset, self.padlen = self.find_offset()
+            log.info("Found format string offset: %d", self.offset)
+
+        self.writes = []
+        self.leaker = MemLeak(self._leaker)
+
+    def leak_stack(self, offset, prefix=""):
+        leak = self.execute_fmt(prefix+"START%%%d$pEND" % offset)
+        try:
+            leak = re.findall(r"START(.*)END", leak, re.MULTILINE | re.DOTALL)[0]
+            leak = int(leak, 16)
+        except ValueError:
+            leak = 0
+        return leak
+
+    def find_offset(self):
+        marker = cyclic(20)
+        for off in range(1,1000):
+            leak = self.leak_stack(off, marker)
+            leak = p32(leak)
+
+            pad = cyclic_find(leak)
+            if pad >= 0 and pad < 20:
+                return off, pad
+        else:
+            log.error("Could not find offset to format string on stack")
+            return None, None
+
+    def _leaker(self, addr):
+        # Hack: elfheaders often start at offset 0 in a page,
+        # but we often can't leak addresses containing null bytes,
+        # and the page below elfheaders is often not mapped.
+        # Thus everything on a page boundry is a "\x7f"
+        # unless it is leaked otherwise.
+        if addr & 0xfff == 0: return "\x7f"
```

Comment:
Alternately, we could assume that we can read from the page before the x000-aligned address *unless* the bytes at +1-3 are 'ELF'.